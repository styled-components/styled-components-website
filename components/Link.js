import styled from 'styled-components';
import UnstyledLink from 'next/link';

import rem from '../utils/rem';
import { red, blmGrey, lightGrey } from '../utils/colors';

export const StyledLink = styled.a`
  display: inline-block;
  color: inherit;
  cursor: pointer;
  padding: ${rem(2)} ${rem(8)};
  margin: ${rem(-2)} ${rem(-8)};

  @media (min-width: ${1000 / 16}em) {
    border-radius: ${rem(3)};

    &:hover {
      background: ${lightGrey};
    }
  }
`;

export const InlineLink = styled.a.attrs((/* props */) => ({
  target: '_blank',
  rel: 'noopener',
}))`
  color: ${(p) => (p['data-white'] ? 'white' : blmGrey)};
  cursor: pointer;
  text-decoration: underline;

  &:hover {
    color: ${(p) => (p['data-white'] ? 'white' : red)};
  }
`;

const Link = ({ ['aria-label']: ariaLabel, children, className, inline, unstyled, white, ...rest }) => {
  let Child = StyledLink;
  if (inline) {
    Child = InlineLink;
  } else if (unstyled) {
    Child = 'a';
  }

  let dataAttrs;
  if (white) {
    dataAttrs = { 'data-white': white };
  }

  return (
    <UnstyledLink {...rest}>
      <Child aria-label={ariaLabel} href={rest.href} className={className} {...dataAttrs}>
        {children}
      </Child>
    </UnstyledLink>
  );
};

export default Link;
