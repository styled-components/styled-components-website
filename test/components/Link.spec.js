import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Link, { StyledLink, InlineLink } from '../../components/Link';

test('Link renders correctly', () => {
  const tree = renderer.create(<Link href="/" />).toJSON();

  expect(tree).toMatchSnapshot();
});

test('StyledLink renders correctly', () => {
  const tree = renderer.create(<StyledLink href="/" />).toJSON();

  expect(tree).toMatchSnapshot();
});

test('InlineLink renders correctly', () => {
  const tree = renderer.create(<InlineLink href="/" />).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Link is white', () => {
  const wrapper = mount(<Link href="/" inline white />);

  expect(wrapper).toHaveStyleRule('color', 'white');
});
