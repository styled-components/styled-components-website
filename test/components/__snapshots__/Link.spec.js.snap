// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineLink renders correctly 1`] = `
.c0 {
  color: rgb(33,33,33);
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0:hover {
  color: #ff5555;
}

<a
  className="c0"
  href="/"
  rel="noopener"
  target="_blank"
/>
`;

exports[`Link renders correctly 1`] = `
.c0 {
  display: inline-block;
  color: inherit;
  cursor: pointer;
  padding: 0.1111111111111111rem 0.4444444444444444rem;
  margin: -0.1111111111111111rem -0.4444444444444444rem;
}

@media (min-width:62.5em) {
  .c0 {
    border-radius: 0.16666666666666666rem;
  }

  .c0:hover {
    background: rgba(20,20,20,0.1);
  }
}

<a
  className="c0"
  href="/"
/>
`;

exports[`StyledLink renders correctly 1`] = `
.c0 {
  display: inline-block;
  color: inherit;
  cursor: pointer;
  padding: 0.1111111111111111rem 0.4444444444444444rem;
  margin: -0.1111111111111111rem -0.4444444444444444rem;
}

@media (min-width:62.5em) {
  .c0 {
    border-radius: 0.16666666666666666rem;
  }

  .c0:hover {
    background: rgba(20,20,20,0.1);
  }
}

<a
  className="c0"
  href="/"
/>
`;
